File: forms/neurodiversity-adjustment-review/front-end-with-svelte/src/lib/engine/next-step-rules.ts
import type {
	NeurodiversityAdjustmentReview,
	WellbeingRiskBand,
	NextStepUrgency,
	FiredRule
} from './types';
import { anyNotWorking } from './utils';

/**
 * Axis D — next-step urgency, plus the target timeframe.
 *
 * First match wins. An escalation, high wellbeing risk, or a failing / changing
 * adjustment auto-raise the urgency regardless of whether the next review is
 * booked; the least-urgent band (`none`) is chosen only when no rule fires.
 */
export function gradeNextStep(
	r: NeurodiversityAdjustmentReview,
	wellbeingRiskBand: WellbeingRiskBand
): {
	nextStepUrgency: NextStepUrgency;
	targetTimeframe: string;
	firedRules: FiredRule[];
} {
	const firedRules: FiredRule[] = [];

	if (r.escalated) {
		firedRules.push({
			ruleId: 'R-NEXT-ESCALATED',
			axis: 'next-step',
			category: 'escalate',
			description: 'Escalation in progress — follow the escalation procedure.'
		});
		return {
			nextStepUrgency: 'escalate',
			targetTimeframe: 'Escalate now',
			firedRules
		};
	}

	if (wellbeingRiskBand === 'high-risk') {
		firedRules.push({
			ruleId: 'R-NEXT-HIGH-RISK',
			axis: 'next-step',
			category: 'adjust-now',
			description: 'High wellbeing risk — act now.'
		});
		return {
			nextStepUrgency: 'adjust-now',
			targetTimeframe: 'Within 2 weeks',
			firedRules
		};
	}

	if (anyNotWorking(r) || r.changesNeeded === true) {
		firedRules.push({
			ruleId: 'R-NEXT-CHANGES',
			axis: 'next-step',
			category: 'adjust-now',
			description: 'A failing adjustment or an agreed change needs action.'
		});
		return {
			nextStepUrgency: 'adjust-now',
			targetTimeframe: 'Within 2 weeks',
			firedRules
		};
	}

	if (r.nextReviewDate.trim() !== '') {
		firedRules.push({
			ruleId: 'R-NEXT-REVIEW-SCHEDULED',
			axis: 'next-step',
			category: 'review-scheduled',
			description: 'Next review is scheduled.'
		});
		return {
			nextStepUrgency: 'review-scheduled',
			targetTimeframe: r.nextReviewDate,
			firedRules
		};
	}

	firedRules.push({
		ruleId: 'R-NEXT-NONE',
		axis: 'next-step',
		category: 'none',
		description: 'No further action scheduled.'
	});
	return {
		nextStepUrgency: 'none',
		targetTimeframe: 'No follow-up scheduled',
		firedRules
	};
}

File: forms/neurodiversity-adjustment-review/back-end-with-loco/src/neurodiversity_adjustment_review/engine/next_step_rules.rs
//! Axis D — next-step urgency, plus the target timeframe.
//!
//! Urgency ladder `none → review-scheduled → adjust-now → escalate`. Exactly
//! one rule fires (first match wins). An escalation, a high wellbeing-risk band,
//! a failing adjustment, or an agreed change drives the most-urgent bands
//! regardless of the review schedule.

use super::types::{FiredRule, NeurodiversityAdjustmentReview, NextStepUrgency, WellbeingRiskBand};
use super::utils::any_not_working;

/// The Axis-D outcome: urgency, target timeframe, and fired rule.
pub struct NextStep {
    /// Next-step urgency.
    pub next_step_urgency: NextStepUrgency,
    /// Target timeframe for the next step.
    pub target_timeframe: String,
    /// Fired rule for the audit trail.
    pub fired_rules: Vec<FiredRule>,
}

/// Grade next-step urgency from the review and the already-computed
/// wellbeing-risk band.
#[must_use]
pub fn grade_next_step(
    r: &NeurodiversityAdjustmentReview,
    wellbeing_risk_band: &WellbeingRiskBand,
) -> NextStep {
    let (urgency, rule_id, category, description): (&str, &str, &str, &str) = if r.escalated {
        (
            "escalate",
            "R-NEXT-ESCALATED",
            "escalated",
            "Escalation in progress — follow the escalation procedure.",
        )
    } else if wellbeing_risk_band == "high-risk" {
        (
            "adjust-now",
            "R-NEXT-HIGH-RISK",
            "high-risk",
            "High wellbeing risk — act now.",
        )
    } else if any_not_working(r) || r.changes_needed {
        (
            "adjust-now",
            "R-NEXT-CHANGES",
            "changes",
            "A failing adjustment or an agreed change needs action.",
        )
    } else if !r.next_review_date.trim().is_empty() {
        (
            "review-scheduled",
            "R-NEXT-REVIEW-SCHEDULED",
            "review-scheduled",
            "Next review is scheduled.",
        )
    } else {
        (
            "none",
            "R-NEXT-NONE",
            "none",
            "No further action scheduled.",
        )
    };

    let target_timeframe = match urgency {
        "escalate" => "Escalate now".to_string(),
        "adjust-now" => "Within 2 weeks".to_string(),
        "review-scheduled" => {
            if r.next_review_date.trim().is_empty() {
                "At the scheduled review".to_string()
            } else {
                r.next_review_date.trim().to_string()
            }
        }
        _ => "No follow-up scheduled".to_string(),
    };

    NextStep {
        next_step_urgency: urgency.to_string(),
        target_timeframe,
        fired_rules: vec![FiredRule {
            rule_id: rule_id.to_string(),
            axis: "next-step".to_string(),
            category: category.to_string(),
            description: description.to_string(),
        }],
    }
}
